weaveworks/weave-gitops
# GitOps Run failing validation on non-Kubernetes yaml files

Closes #3491 

Changes:
- Added the `sourceignore` package with functions to create the `.sourceignore` file (called in the new `InitializeRootDir` function) and read ignore patterns from it.

- Add default ignore patterns.

- Added tests for the `sourceignore` package.

- Added ignoring files according to ignore patterns, loaded from .sourceignore, during YAML validation.

- Added watching for changes in the `.sourceignore` file to trigger reconciliation of resources and re-validation of YAML files.


Notes:
- Error when creating the .sourceignore file and the file already exists (if this type of error `sourceignore.ErrIgnoreFileExists` is returned from the `CreateIgnoreFile` function) will be ignored and the context will not be cancelled.
- The error when reading ignore files from the ignore patterns before validation will be logged as a warning to the console.
- The logic of how default ignore patterns are named and created a borrowed from FluxCD(which, AFAIR, uses just VC patterns for creating a matcher separately somewhere, that's why the separate `VCSPatterns` function, but if you prefer I could merge it with the default patterns, but I thought we could use smth. like that in the future).

Testing:
- Please run GitOps Run with either `.` as the target directory or a subdirectory as the target dir
- create some "invalid" non-Kubenetes YAMLs (with .yaml or .yml extensions) in the target directory
- check that the validation errors are displayed for those "invalid" YAML and add patterns to ignore them to the newly created `.sourceignore` file.
